marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/mappingUILib.sjs

```diff
@@ -0,0 +1,126 @@
+        }
+      }
+      else if(!entitiesNotToExpand.includes(relatedEntityName)){
+        let nextEntityResponseObject ={};
+        propertyPath += "." + entityPropertyName;
+        nextEntityResponseObject.entityMappingId = relatedEntityName + ":" + propertyPath;
+        nextEntityResponseObject.entityType = relatedEntityName;
+        nextEntityResponseObject.mappingTitle = relatedEntityName + "(" + entityName + " " +entityPropertyName + ")";
+        if(!entityResponseObject.relatedEntityMappings){
+          entityResponseObject.relatedEntityMappings = [];
+        }
+        entityResponseObject.relatedEntityMappings.push({"mappingLinkText": entityPropertyName + " " + relatedEntityName, "entityMappingId": nextEntityResponseObject.entityMappingId})
+        entitiesNotToExpand.push(relatedEntityName)
+        let relatedEntityProperties = expandStructuredProperties(getEntityModel(relatedEntityName), relatedEntityName);
+        getAllRelatedEntities(relatedEntityName, targetEntityName, relatedEntityProperties , entitiesNotToExpand ,  nextEntityResponseObject, relatedEntityName);
+        entitiesNotToExpand.pop();
+      }
+    }
+
+    if(!entityResponseObject.entityModel){
+      entityResponseObject.entityModel =  expandedEntityModel;
+      responseArray.push(entityResponseObject);
+    }
+    if (entityPropertyValue["subProperties"]) {
+      getAllRelatedEntities(entityName, targetEntityName, entityPropertyValue["subProperties"], entitiesNotToExpand, entityResponseObject, propertyPath+ "." + entityPropertyName);
+      propertyPath = propertyPath.lastIndexOf(".") != -1 ? propertyPath.substring(0, propertyPath.lastIndexOf(".")) : propertyPath;
+    }
+  }
+}
+
+function getEntityModel(entityName){
+  return modelArray.find(model => model.info.title == entityName);
+}
+
+function expandStructuredProperties(entityModel, entityName) {
+  let entityProperties = entityModel.definitions[entityName].properties;
+
+  for(let entityPropertyName in entityProperties){
+    let entityPropertyValue = entityProperties[entityPropertyName];
+    if(entityPropertyValue["$ref"] || (entityPropertyValue["items"] && entityPropertyValue["items"]["$ref"])){
+      let ref = entityPropertyValue["$ref"] ? entityPropertyValue["$ref"] : entityPropertyValue["items"]["$ref"];
+      if(ref.startsWith("#/")){
+        let subEntityName = ref.substring(ref.lastIndexOf('/') + 1);
+        entityPropertyValue["subProperties"] = expandStructuredProperties(entityModel, subEntityName);
+      }
+      else{
+        //Remove external references as they shouldn't show up in mapping
+        delete entityProperties[entityPropertyName];
+      }
+    }
+    else{
+      entityProperties[entityPropertyName] = entityPropertyValue;
+    }
+  }
+  return entityProperties;
+}
+
+
+function getEntityModelsWithPropertyThatRefersToTargetEntity(targetEntityModel, targetEntityName){
```

Comment:
Let's use "find" here instead of "get" to make it clear that a query is being performed.